apptools/undo/traitsui/trait_set_command.py

```diff
@@ -0,0 +1,82 @@
+#
+# (C) Copyright 2015 Enthought, Inc., Austin, TX
+# All right reserved.
+#
+# This file is open source software distributed according to the terms in
+# LICENSE.txt
+#
+"""
+TraitSetCommand
+===============
+
+This module provides a simple command which sets a trait value on ``do`` and
+``redo``, resets it to the previous value on ``undo``, and merges multiple
+commands on the same attribute on the same object.
+
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+# Enthought library imports
+from traits.api import Any, Str
+
+# Local library imports
+from apptools.undo.api import AbstractCommand
+
+
+class TraitSetCommand(AbstractCommand):
+    """ Command subclass which simply sets a trait on an object
+
+    This is somewhat simplistic: it assumes that nothing else is going to
+    hold on to and mutate the old value unexpectedly, which may break down
+    with things like NumPy arrays.  The main purposes is to generate commands
+    from the UndoHandler, so we can be reasonably comfortable that the values
+    are being handled sanely.
+
+    """
+
+    #-------------------------------------------------------------------------
+    # 'TraitSetCommand' interface
+    #-------------------------------------------------------------------------
+
+    #: The name of the data trait we are acting on.
+    trait_name = Str
+
+    #: The new value that the trait is being set to
+    value = Any
+
+    #: The old value of the trait
+    _old_value = Any
+
+    #-------------------------------------------------------------------------
+    # 'ICommand' interface
+    #-------------------------------------------------------------------------
+
+    #: The user-visible name of the command
+    name = Str
+
+    def do(self):
+        # we assume that copy or deep copy isn't required
+        self._old_value = getattr(self.data, self.trait_name)
+        self.redo()
+
+    def redo(self):
+        setattr(self.data, self.trait_name, self.value)
+
+    def undo(self):
+        setattr(self.data, self.trait_name, self._old_value)
+
+    def merge(self, other):
+        if (isinstance(other, self.__class__) and self.data is other.data and
+                self.trait_name == other.trait_name):
+            self.value = other.value
+            return True
```

Comment:
I know, but this just indicates that this is not the right place to shim in undo functionality. It needs to be closer to the UI control since only some UI controls have this issue (admittedly, very important ones). The implementation of the `Editor` needs to be responsible for determining when estimating when a user has completed a single logical action even if it sends live trait updates throughout. E.g. a slider starts the logical action on button-down and completes it with button-up; the undo command should only be added to the stack on button-up. Similarly, a text field would "logically complete" for undo purposes on enter-pressed or focus-out (even if the `Editor` was configured to update the model live). Estimating logical completeness of the user's action based on the timing of the event stream is much less reliable, but might be workable (the tests will be a nightmare, of course). One option would be to allow trait or `Item` metadata that can inform the `UndoHandler` whether it should merge adjacent changes or not. So `TextEditors` and `RangeEditors` could merge while `EnumEditors` wouldn't, at the expense of making the coder be explicit about it.

Yes, the complexity of the code needs to increase and may need changes to the `Editors` themselves. It's a complex problem, and the information needed to handle it has been stripped out at this point of the framework. As it stands, I think this set of undocumented limitations makes this an attractive nuisance given where this code is located and how it is named. I think the user experience will be surprising in ways that diminish confidence in the user's perception of how safe it is to explore the UI, which is the point of building in undo functionality.